url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/RTree.java

```diff
@@ -0,0 +1,1674 @@
+					// provided (one that stores any node in consecutive pages) performance will improve substantially,
+					// since disk accesses will be mostly sequential. On the other hand, the index will need to
+					// manipulate the data, resulting in larger overhead. If you use a main memory storage manager,
+					// storing the data externally is highly recommended (clustering has no effect).
+					// A clustered storage manager is NOT provided yet.
+					// Also you will have to take care of converting you data to and from binary format, since only
+					// array of bytes can be inserted in the index (see RTree::Node::load and RTree::Node::store for
+					// an example of how to do that).
+
+					//tree.insertData(data.getBytes(), r, id);
+
+					tree.insertData(null, r, id);
+					id ++;
+					// example of passing a null pointer as the associated data.
+				} else if (op == 2) {
+					//query
+
+
+					f1[0] = x1;
+					f1[1] = y1;
+					f1[2] = Math.sqrt(z1);
+					f2[0] = x1;
+					f2[1] = y1;
+					f2[2] = Math.sqrt(z1);
+					
+					Region r = new Region(f1, f2);
+
+					//MyVisitor vis = new MyVisitor();
+					//SecurityVisitor vis = new SecurityVisitor(r);
+
+					if (args[3].equals("intersection")) {
+						//tree.intersectionQuery(r, vis);
+						// this will find all data that intersect with the query range.
+					} else if (args[3].equals("10NN")) {
+						//Point p = new Point(f1);
+						//tree.nearestNeighborQuery(10, p, vis);
+						// this will find the 10 nearest neighbors.
+					} else {
+						println("Unknown query type.");
+						System.exit(-1);
+					}
+				}
+
+				if (count % 1000 == 0) {
+					System.out.print(".");
+					if(count % 10000 == 0)println(count + "");
+				}
+
+				count++;
+				line = lr.readLine();
+			}
+//			System.out.println("Total id :\t" + id);
+			long end = System.currentTimeMillis();
+
+			println("Operations: " + count);
+			println(tree.toString());
+			println("Minutes: " + (end - start) / 1000.0f / 60.0f);
+
+			// since we created a new RTree, the PropertySet that was used to initialize the structure
+			// now contains the IndexIdentifier property, which can be used later to reuse the index.
+			// (Remember that multiple indices may reside in the same storage manager at the same time
+			//  and every one is accessed using its unique IndexIdentifier).
+			Integer indexID = (Integer) ps2.getProperty("IndexIdentifier");
+			println("Index ID: " + indexID);
+
+			boolean ret = tree.isIndexValid();
+			if (ret == false) {
+				println("Structure is INVALID!");
+			}
+
+			// flush all pending changes to persistent storage (needed since Java might not call finalize when JVM exits).
+			tree.flush();
+			return (RTree) tree;
+		} catch (Exception e) {
+			e.printStackTrace();
+			return null;
+		}
+	}
+
+	public void loadIndexOfPoints(){
+		try {
+			DataInputStream dis = new DataInputStream(new BufferedInputStream(new FileInputStream(new File(destFileNamePoint + ".idx"))));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dis*. The resource is referenced at lines 357, 360, 361. The resource is closed at line 363. There are other execution paths that do not contain closure statements, for example, when DataInputStream.readLong throws an exception. Ensure *dis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).